### PR TITLE
Estimate test

### PR DESCRIPTION
**Pull request by qishuny**
estimate using imu data and comment out quadKD code

### CODE REVIEW

File: ekf_estimator/include/ekf_estimator/ekf_estimator.h [hunk 1]
```diff
@@ -28,6 +32,13 @@ class EKFEstimator {
   void spin();
 
 private:
+
+  /**
+   * @brief Callback function to handle new ground truth data
+   * @param[in] ground_truth_msg quad_msgs<RobotState>
```
Comment (jcnorby) [inline ekf_estimator/include/ekf_estimator/ekf_estimator.h:38]: `  * @param[in] msg ground_truth_msg quad_msgs<RobotState>`

File: ekf_estimator/include/ekf_estimator/ekf_estimator.h [hunk 2]
```diff
@@ -52,6 +63,13 @@ class EKFEstimator {
    */
   quad_msgs::RobotState updateStep();
 
+
+  /// Subscriber for joint encoder messages
+  ros::Subscriber state_ground_truth_pub_;
```
Comment (jcnorby) [inline ekf_estimator/include/ekf_estimator/ekf_estimator.h:68]: Comment, type, and name are different

File: ekf_estimator/include/ekf_estimator/ekf_estimator.h [hunk 3]
```diff
@@ -85,5 +103,33 @@ class EKFEstimator {
...
+  // IMU bias in acc z
+  double bias_z_;
+  
+  // QuadKD class
+  // std::shared_ptr<quad_utils::QuadKD>quadKD_;
+
+ 
+
+
+
```
Comment (jcnorby) [inline ekf_estimator/include/ekf_estimator/ekf_estimator.h:133]: Can clear these lines

File: ekf_estimator/src/ekf_estimator.cpp [hunk 4]
```diff
@@ -1,22 +1,42 @@
 #include "ekf_estimator/ekf_estimator.h"
 
+
+const int numofStates = 6;
+const int num_feet_ = 4;
```
Comment (jcnorby) [inline ekf_estimator/src/ekf_estimator.cpp:5]: You can make these static const ints that are member variables of the class so they aren't in the global scope.

File: ekf_estimator/src/ekf_estimator.cpp [hunk 5]
```diff
@@ -1,22 +1,42 @@
...
 EKFEstimator::EKFEstimator(ros::NodeHandle nh) {
   nh_ = nh;
 
   // Load rosparams from parameter server
   std::string joint_encoder_topic, imu_topic, contact_topic, state_estimate_topic;
-  nh.param<std::string>("topics/joint_encoder", joint_encoder_topic, "/joint_encoder");
-  nh.param<std::string>("topics/imu", imu_topic, "/imu");
+  nh.param<std::string>("topics/mcu/state/jointURDF", joint_encoder_topic, "/mcu/state/jointURDF");
+  nh.param<std::string>("topics/mcu/state/imu", imu_topic, "/mcu/state/imu");
```
Comment (jcnorby) [inline ekf_estimator/src/ekf_estimator.cpp:12]: Is there a reason for this change? I think we want to load them from the topics/joint_encoder and topics/imu parameters (although this won't matter once this is all implemented within RobotDriver).

File: ekf_estimator/src/ekf_estimator.cpp [hunk 6]
```diff
@@ -38,56 +58,141 @@ quad_msgs::RobotState EKFEstimator::updateStep() {
...
+      0.308681,
+      0,
+      0,
+      0;
+  x = x0;
+  last_x = x;
+
+  // std::cout << "estimate state x" << x << std::endl;
+  std::cout << "bias x is" << bias_x_ << std::endl;
+  std::cout << "bias y is" << bias_y_ << std::endl;
```
Comment (jcnorby) [inline ekf_estimator/src/ekf_estimator.cpp:195]: Clear these out for now

File: quad_utils/data/EKF_tests_2021_12_3/.~lock.test2.csv# [hunk 7]
```diff
@@ -0,0 +1 @@
+,qishuny,qishuny-5510,04.02.2022 17:11,file:///home/qishuny/.config/libreoffice/4;
```
Comment (jcnorby) [inline quad_utils/data/EKF_tests_2021_12_3/.~lock.test2.csv#:1]: Add this to .gitignore

File: quad_utils/launch/test_ekf.launch [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        <!-- Launch the nodes to publish all the robot tfs and visualize in rviz -->
+        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
+            <remap from="joint_states" to="/visualization/joint_states/ground_truth"/>
+            <param name="tf_prefix" value="ground_truth"/>
+        </node>
+    </group>
+
+    <group ns="/estimate">
+        <!-- Launch the node to publish state estimate messages-->
+        <node name="ekf_estimator" pkg="ekf_estimator" type="ekf_estimator_node" output="screen"/>
```
Comment (jcnorby) [inline quad_utils/launch/test_ekf.launch:25]: This is likely why you needed to add the "/" in front of "robot_description" - if launching ekf estimator under a namespace, the resulting node would be /estimate/ekf_estimator, and looking for a parameter named "robot_description" would result in looking for "/estimate/robot_description". The leading "/" clarifies to look in the global namespace.

I think the best fix for this is to just launch the estimator outside of this namespace (either directly or ideally by including the estimation.launch file).

File: quad_utils/src/quad_kd.cpp [hunk 9]
```diff
@@ -7,7 +7,7 @@ Eigen::IOFormat CleanFmt(4, 0, ", ", "\n", "[", "]");
 QuadKD::QuadKD()
 {
   std::string robot_description_string;
-  if (!ros::param::get("robot_description", robot_description_string))
+  if (!ros::param::get("/robot_description", robot_description_string))
```
Comment (jcnorby) [inline quad_utils/src/quad_kd.cpp:10]: See above comment, although it might be fine to leave this as checking the global namespace for now.

File: quad_utils/src/rviz_interface.cpp [hunk 10]
```diff
@@ -429,6 +428,7 @@ void RVizInterface::robotStateCallback(
     transformStamped.child_frame_id = "/estimate/body";
     estimate_base_tf_br_.sendTransform(transformStamped);
     estimate_joint_states_viz_pub_.publish(joint_msg);
+    // std::cout <<" published estimated joint_msg" <<std::endl;
```
Comment (jcnorby) [inline quad_utils/src/rviz_interface.cpp:431]: You can revert these changes

File: quad_utils/rviz/example_with_terrain.rviz [hunk 11]
```diff
@@ -3,9 +3,11 @@ Panels:
     Help Height: 0
```
Comment (jcnorby) [inline quad_utils/rviz/example_with_terrain.rviz:3]: Revert this file for now since we aren't ready to fully support the state estimate yet.

File: ekf_estimator/src/ekf_estimator.cpp (hunk 5 above)
Comment (qishuny) [reply ekf_estimator/src/ekf_estimator.cpp:12]: Yes. Currently, I can only get the imu and encoder data from the bag file and they are published to "topics/mcu/state/jointURDF" and "topics/mcu/state/imu". But yeah, it won't matter once this is implemented within RobotDriver.

File: ekf_estimator/include/ekf_estimator/ekf_estimator.h [hunk 12]
```diff
@@ -63,8 +63,11 @@ class EKFEstimator {
    */
   quad_msgs::RobotState updateStep();
 
+  
+  const int numofStates = 6;
+  const int num_feet_ = 4;
 
-  /// Subscriber for joint encoder messages
+  /// Subscriber for ground_truth RobotState messages
   ros::Subscriber state_ground_truth_pub_;
```
Comment (jcnorby) [inline ekf_estimator/include/ekf_estimator/ekf_estimator.h:71]: Should this be state_ground_truth_sub_? or pub_?

File: quad_utils/rviz/example_with_terrain.rviz [hunk 13]
```diff
@@ -1,6 +1,5 @@
 Panels:
   - Class: rviz/Displays
-    Help Height: 0
```
Comment (jcnorby) [inline quad_utils/rviz/example_with_terrain.rviz:3]: I still see the (unpublished) Estimated State in the RViz view, can we revert that so it doesn't show up by default?

**Review by jcnorby — CHANGES_REQUESTED** [review]

**Review by jcnorby — APPROVED** [review]
